### PR TITLE
Tpetra: "New RTI" (Reduction Transform Interface), Part 1 (for_each, transform)

### PR DESCRIPTION
**Pull request by mhoemmen**
@trilinos/tpetra 

## Description

Tpetra once had something called the Reduction Transform Interface (RTI).  It let users apply a function to each entry of one or two (or three, I think) `Tpetra::Vector`s, and then possibly end with a reduction over the output Vector.

This commit starts anew with RTI, by adding the following features:

  1. `withLocalAccess`: Access the local data of a `Tpetra::)Multi)Vector` in a given memory space -- a declarative replacement for DualView-style functions
  2. `for_each`: Works like `std::for_each`, but for `Tpetra::(Multi)Vector`
  3. `transform`: Works like unary or binary `std::transform`, but for `Tpetra::(Multi)Vector`

(2) and (3) use (1).  All of these have extension points so that they can be adapted for other objects, such as CrsGraph or CrsMatrix.

## Motivation and Context

Users want to apply functions entrywise to a Vector or MultiVector.  They want to get at the local data, and they want Tpetra to handle the sync and modify flag stuff for them.  Tpetra wants the power to do things like lazy allocation (see #333) that go beyond what `Kokkos::DualView` allows.  Tpetra wants to protect users from holding on to owning `Kokkos::View` that really belong to Tpetra objects, because that can cause bugs (e.g., if the mesh that produced the linear system has changed) and unnecessary memory usage.

## Next steps

  1. Add `transform_reduce` (the "R" in "RTI")
  2. Think about how to do kernel fusion with a mixture of Vector and CrsMatrix (or at least provide a "residual" interface)

## Related Issues

* Closes: #3471 
* Related to #333, #364, #381, #415, #768, #1424, #1896

### CODE REVIEW

File: packages/tpetra/core/test/MultiVector/ForEach.cpp [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+    reduceAll<int, int> (*comm, REDUCE_MIN, lclSuccess, outArg (gblSuccess));
+    TEST_ASSERT( gblSuccess == 1 );
+    if (gblSuccess != 1) {
+      out << "Returning early" << endl;
+      return;
+    }
+
+    out << "X.sync_device(); X.putScalar(666.0);" << endl;
+    X.sync_device ();
+    X.putScalar (666.0);
```
Comment (csiefer2) [inline packages/tpetra/core/test/MultiVector/ForEach.cpp]: I see what you did there...

Comment (mhoemmen) [reply packages/tpetra/core/test/MultiVector/ForEach.cpp]: "It's a solar number.  You may call me 'Little Sunshine.'"

**Comment by mhoemmen** [comment]
@nmhamster Check this out.  If `withLocalAccess` were the only way to get at a Tpetra object's local data, then we could totally detach Tpetra from MPI and let the parallel distribution change whenever it wants.

**Comment by mhoemmen** [comment]
Weird Intel build errors:
```
/scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_intel_17.0.1/Trilinos/packages/tpetra/core/src/Tpetra_for_each_MultiVector.hpp(234): error: a built-in binary operator applied to a scoped enumeration requires two operands of the same type
        static_assert (argsType != EMultiVectorForEachFuncArgs::ERROR,
```

**Comment by mhoemmen** [comment]
@trilinos/tpetra developers reviewed this on Wednesday, 08 May 2019.  We decided the following:

1. withLocalAccess is fine as is.
2. Remove `value&, row index, column index` and `value&, row index` overloads of `for_each.

My new commits implement these changes.  Tests pass with CUDA 9.2 + GCC 7.2.0 and with Intel 17.0.1 + OpenMP (with `OMP_NUM_THREADS=4`).  I also redesigned GetNonowningLocalObject to make specialization for custom types easier, and added tests that exercise with withLocalAccess for a custom type (to make sure that specialization for custom types actually works).

**Comment by mhoemmen** [comment]
@trilinos/framework Those are old PR test results.  Not sure why they haven't updated.

**Comment by mhoemmen** [comment]
Not sure what's up w/ the SERIAL build.  I tested CUDA 9.2 and Intel 17.0.1 + OpenMP by hand.  I'll take another look.

**Comment by mhoemmen** [comment]
@trilinos/framework Dashboard link is broken: "Error: CDash cannot connect to the database."

**Comment by mhoemmen** [comment]
I set up a GCC 4.9.3 "serial" (MPI disabled) OpenMP build, and got a build error:
```
.../packages/tpetra/core/src/Tpetra_for_each_MultiVector.hpp:189:38: error: invalid type argument of unary ‘*’ (have ‘auto’)
                 decltype (readWrite (*X_j).on (memSpace))>;
```
This could be a compiler bug, since other compilers have no problem deducing the type of `X_j` here.  I will try to work around by assigning `*X_j` to `Tpetra::Vector&` explicitly.

**Comment by mhoemmen** [comment]
@trilinos/tpetra My work-around for GCC 4.9.3 issues was successful.  Please review if you have a chance.  Thanks!

**Comment by prwolfe** [comment]
> @trilinos/framework Dashboard link is broken: "Error: CDash cannot connect to the database." Here is the link: https://testing.sandia.gov/cdash/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=PR-5115-test&field2=buildstarttime&compare2=84&value2=NOW

Must have been transient as it works for me at this point.

**Comment by mhoemmen** [comment]
@prwolfe wrote:
> Must have been transient as it works for me at this point.

Cool, glad it works now :-D

**Comment by mhoemmen** [comment]
@csiefer2 U R THE MAN

**Pull request merged**